blacklistfilter/ipblacklistfilter.cpp

```diff
@@ -0,0 +1,740 @@
+    } else {
+        ur_set(ur_det, detected, F_DST_BLACKLIST, 0x0);
+    }
+
+    if (marked) {
+        // IP was found
+        return BLACKLISTED;
+    }
+    return ADDR_CLEAR;
+}
+
+
+
+int main (int argc, char **argv)
+{
+    int retval = 0;
+    int send_terminating_unirec = 1;
+
+    // Set default files names
+    char *userFile = (char*) SYSCONFDIR "/blacklistfilter/ipdetect_config.xml";
+
+    // For use with prefixes
+    black_list_t v4_list;
+    black_list_t v6_list;
+
+    // Mask array for prefixes
+    ipv4_mask_map_t v4_masks;
+    ipv6_mask_map_t v6_masks;
+    create_v4_mask_map(v4_masks);
+    create_v6_mask_map(v6_masks);
+
+    // TRAP initialization
+    //TRAP_DEFAULT_INITIALIZATION(argc, argv, module_info);
+    INIT_MODULE_INFO_STRUCT(MODULE_BASIC_INFO, MODULE_PARAMS)
+
+    trap_ifc_spec_t ifc_spec;
+    int ret = trap_parse_params(&argc, argv, &ifc_spec);
+    if (ret != TRAP_E_OK) {
+        if (ret == TRAP_E_HELP) {
+            trap_print_help(module_info);
+            return 0;
+        }
+        trap_free_ifc_spec(ifc_spec);
+        fprintf(stderr, "ERROR in parsing of parameters for TRAP: %s\n", trap_last_error_msg);
+        return 1;
+    }
+    ret = trap_init(module_info, ifc_spec);
+    if (ret != TRAP_E_OK) {
+        trap_free_ifc_spec(ifc_spec);
+        fprintf(stderr, "ERROR in TRAP initialization: %s\n", trap_last_error_msg);
+        return 1;
+    }
+
+    trap_ifcctl(TRAPIFC_OUTPUT, 0, TRAPCTL_SETTIMEOUT, TRAP_HALFWAIT);
+    trap_free_ifc_spec(ifc_spec);
+
+    // UniRec templates for recieving data and reporting blacklisted IPs
+    char *errstr = NULL;
+    ur_template_t *templ = ur_create_input_template(0, "SRC_IP,DST_IP,SRC_PORT,DST_PORT,PROTOCOL,PACKETS,BYTES,TIME_FIRST,TIME_LAST,TCP_FLAGS,LINK_BIT_FIELD,DIR_BIT_FIELD,TOS,TTL", &errstr);
+
+    if (templ == NULL) {
+        std::cerr << "Error: Invalid UniRec specifier." << std::endl;
+        if(errstr != NULL){
+            fprintf(stderr, "%s\n", errstr);
+            free(errstr);
+        }
+        trap_finalize();
+        return EXIT_FAILURE;
+    }
+
+    ur_template_t *tmpl_det = ur_create_output_template(0, "SRC_IP,DST_IP,SRC_PORT,DST_PORT,PROTOCOL,PACKETS,BYTES,TIME_FIRST,TIME_LAST,TCP_FLAGS,LINK_BIT_FIELD,DIR_BIT_FIELD,TOS,TTL,SRC_BLACKLIST,DST_BLACKLIST", &errstr);
+    if (tmpl_det == NULL) {
+        std::cerr << "Error: Invalid UniRec specifier." << std::endl;
+        if(errstr != NULL){
+            fprintf(stderr, "%s\n", errstr);
+            free(errstr);
+        }
+        trap_finalize();
+        ur_free_template(templ);
+        return EXIT_FAILURE;
+    }
+
+    // Create detection record
+    void *detection = NULL;
+    detection = ur_create_record(tmpl_det, 0);
+    if (detection == NULL) {
+        std::cerr << "ERROR: No memory available for detection report. Unable to continue." << std::endl;
+        ur_free_template(templ);
+        ur_free_template(tmpl_det);
+        return EXIT_FAILURE;
```

Comment:
Should FREE_MODULE_INFO_STRUCT here too. Probably would be good to create a labels and goto statements for error and success handling to avoid repetitive code.